Sources/Fullscreen/FullscreenGallery/FullscreenGalleryViewController.swift

```diff
@@ -0,0 +1,508 @@
+        return vc
+    }
+}
+
+// MARK: - UIPageViewControllerDelegate
+extension FullscreenGalleryViewController: UIPageViewControllerDelegate {
+    public func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
+        guard let imageVc = pageViewController.viewControllers?.first as? FullscreenImageViewController else {
+            return
+        }
+
+        currentImageIndex = imageVc.imageIndex
+        galleryDelegate?.fullscreenGalleryViewController(self, didSelectImageAtIndex: currentImageIndex)
+
+        setCaptionLabel(index: currentImageIndex)
+        previewView.scrollToItem(atIndex: currentImageIndex, animated: true)
+    }
+
+    public func pageViewController(_ pageViewController: UIPageViewController, willTransitionTo pendingViewControllers: [UIViewController]) {
+        pendingViewControllers.forEach({ vc in
+            guard let imageVc = vc as? FullscreenImageViewController else { return }
+
+            imageVc.updateLayout(withPreviewViewVisible: previewViewVisible)
+        })
+    }
+}
+
+// MARK: - UIGesture
+extension FullscreenGalleryViewController: UIGestureRecognizerDelegate {
+    public func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
+        /// gestureRecognizer(_:shouldRequireFailureOf:) is not able to properly prevent the recognizer from triggering
+        /// on touches inside the GalleryPreviewView, so we need to explicitly make sure that the touch is not inside of it.
+        let locationInPreview = touch.location(in: previewView)
+        return !previewView.bounds.contains(locationInPreview)
+    }
+
+    public func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRequireFailureOf otherGestureRecognizer: UIGestureRecognizer) -> Bool {
+        /// Give priority to all other gesture recognizers
+        return true
+    }
+}
+
+// MARK: - FullscreenImageViewControllerDataSource
+extension FullscreenGalleryViewController: FullscreenImageViewControllerDataSource {
+    func loadImage(forImageViewController vc: FullscreenImageViewController, dataCallback: @escaping (UIImage?) -> Void) {
+        guard let galleryDataSource = galleryDataSource else {
+            dataCallback(nil)
+            return
+        }
+
+        let url = viewModel.imageUrls[vc.imageIndex]
+        let imageWidth = min(view.bounds.width, view.bounds.height)
+
+        galleryDataSource.fullscreenGalleryViewController(self, imageForUrlString: url, width: imageWidth, completionHandler: { (_, image, _) in
+            dataCallback(image)
+        })
+    }
+
+    func title(forImageViewController vc: FullscreenImageViewController) -> String? {
+        if vc.imageIndex < 0 || vc.imageIndex >= viewModel.imageCaptions.count {
+            return nil
+        }
+
+        return viewModel.imageCaptions[vc.imageIndex]
```

Comment:
I see you use this alot in your code. We have a custom subscript in `ios-app` which you could probably use here as well.

```swift
return viewModel.imageCaptions[safe: vc.imageIndex]
```

The extension:
```swift
private extension Array {
    // Returns an Optional that will be nil if index < count
    public subscript(safe index: Int) -> Element? {
        return indices.contains(index) ? self[index] : .none
    }
}
```